core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3322,49 +3363,69 @@ private SyncResult syncInodeMetadata(RpcContext rpcContext, LockedInodePath inod
         }
       }
       if (syncPlan.toLoadMetadata()) {
-        loadMetadata = true;
+        AlluxioURI mountUri = new AlluxioURI(mMountTable.getMountPoint(inodePath.getUri()));
+        // We only sync up to the mount point or the originalPath, whichever is lower in the tree
+        if (PathUtils.hasPrefix(originalPath.getUri().getPath(), mountUri.getPath())) {
+          pathsToLoad.add(originalPath.getUri().getPath());
+        } else {
+          pathsToLoad.add(mountUri.getPath());
+        }
       }
       if (syncPlan.toSyncChildren() && inode instanceof InodeDirectory
           && syncDescendantType != DescendantType.NONE) {
         UfsStatus[] listStatus = ufs.listStatus(ufsUri.toString());
+
+        InodeDirectory inodeDir = (InodeDirectory) inode;
+        // maps children name to inode
+        Map<String, Inode<?>> inodeChildren = new HashMap<>();
+        for (Inode<?> child : inodeDir.getChildren()) {
+          inodeChildren.put(child.getName(), child);
+        }
+
+        // if the alluxio path contains a mount point, we need to keep going to find it
         if (listStatus != null) {
-          InodeDirectory inodeDir = (InodeDirectory) inode;
-          // maps children name to inode
-          Map<String, Inode<?>> inodeChildren = new HashMap<>();
-          for (Inode<?> child : inodeDir.getChildren()) {
-            inodeChildren.put(child.getName(), child);
-          }
 
           for (UfsStatus ufsChildStatus : listStatus) {
             if (!inodeChildren.containsKey(ufsChildStatus.getName()) && !PathUtils
                 .isTemporaryFileName(ufsChildStatus.getName())) {
               // Ufs child exists, but Alluxio child does not. Must load metadata.
-              loadMetadata = true;
+              AlluxioURI mountUri = new AlluxioURI(mMountTable.getMountPoint(inodePath.getUri()));
+              // We only sync up to the mount point or the originalPath
+              if (PathUtils.hasPrefix(originalPath.getUri().getPath(), mountUri.getPath())) {
+                pathsToLoad.add(originalPath.getUri().getPath());
+              } else {
+                pathsToLoad.add(mountUri.getPath());
+              }
               break;
             }
           }
+        }
 
+        if (listStatus != null || containsMountPoint) {
```

Comment:
I don't think we have to check for `listStatus != null`?

Also, why do we need to check for `containsMountPoint`? This code block will be traversing the inode children (recursively), so it should eventually find any mount point (if recursive was enabled). Also, even if there is no mount point, a recursive sync should continue the traversal.

Reply:
done